src/equations/compressible_navier_stokes_2d.jl

```diff
@@ -0,0 +1,239 @@
+@doc raw"""
+    CompressibleNavierStokesEquations2D(gamma,
+                                        Re,
+                                        Pr,
+                                        Ma_inf,
+                                        kappa,
+                                        equations)
+
+`CompressibleNavierStokesEquations2D` represents the diffusion (i.e. parabolic) terms applied
+to mass, momenta, and total energy together with the advective from
+the `CompressibleEulerEquations2D`.
+
+gamma: adiabatic constant,
+Re: Reynolds number,
+Pr: Prandtl number,
+Ma_inf: free-stream Mach number
+kappa: thermal diffusivity for Fick's law
+
+For the particular scaling the vicosity is set internally to be μ = 1/Re.
+Further, the nondimensionalization takes the density-temperature-sound speed as
+the principle quantities such that
+rho_inf = 1.0
+T_ref = 1.0
+c_inf = 1.0
+p_inf = 1.0 / gamma
+u_inf = Ma_inf
+R = 1.0 / gamma
+
+Other normalization strategies exist, see the reference below for details.
+- Marc Montagnac (2013)
+  Variable Normalization (nondimensionalization and scaling) for Navier-Stokes
+  equations: a practical guide
+  [CERFACS Technical report](https://www.cerfacs.fr/~montagna/TR-CFD-13-77.pdf)
+The scaling used herein is Section 4.5 of the reference.
+
+In two spatial dimensions we require gradients for three quantities, e.g.,
+primitive quantities
+  grad(u), grad(v), and grad(T)
+or the entropy variables
+  grad(w_2), grad(w_3), grad(w_4)
+where
+  w_2 = rho v_1 / p, w_3 = rho v_2 / p, w_4 = -rho / p
+"""
+# TODO:
+# 1) For now I save gamma and inv(gamma-1) again, but we could potentially reuse them from
+#    the Euler equations
+# 2) Add more here and probably some equations
+struct CompressibleNavierStokesEquations2D{RealT<:Real, E} <: AbstractCompressibleNavierStokesEquations{2, 3}
```

Comment:
Shouldn't this have 4 variables?
```suggestion
struct CompressibleNavierStokesEquations2D{RealT<:Real, E} <: AbstractCompressibleNavierStokesEquations{2, 4}
```

Reply:
This is where I was getting confused with how things were structured. Somewhere we need to keep track of how many gradient variables we will use. If this also tracks the same four variables as Euler would this be 7 (4 cons. vars. and 3 grad. vars.)? My thinking with this is because equations is passed around as a tuple the advective equations has the number of cons vars and the parabolic equations would depend on these variables but includes an additional number for the grad. vars.

Reply:
My impression was that the number of variables in the parabolic equation should correspond to the number of hyperbolic variables. Currently, we just compute gradients of every variable (possibly after a `cons2prim` transformation or something similar), so we don't actually track the number of gradient variables explicitly. For something like CNS with equation-specific optimizations (i.e., avoid computing density gradients and use symmetry of the stress tensor) we would have to specialize for efficiency.

Reply:
That is reasonable to have the number of variables match the advective `equations` number of variables. We could maybe put an `nvars_grad` inside the CNS struct such that the cache creation can declare the gradient arrays to have the correct sizes. In general, I think we need to have the flexibility that the number of conservative variables and the number of gradient variables can be different.

Reply:
Agreed on needing differing cons/grad variables. 

Adding `nvars_grad` as an additional type parameter would help. I was thinking we'd still need to specialize gradient computations based on the equation, but maybe we could do something with specifying `varnames` like we do for shock capturing? 

The other thing I don't know how to account for is symmetry in the stress tensor (e.g., `tau_12 = tau_21 = 0.5 * (dv1dy + dv2dx)`.  My first impression is that we can avoid optimizing based on this, since it mostly saves memory but doesn't reduce the number of operations (we still have to compute `dv1dy` and `dv2dx`, we can just avoid storing them separately). What do you think?

Reply:
For now I think we can still store `dv1dy` and `dv2dx` because for certain types of boundary conditions we might need them anyway. In the current implementation we could exploit the symmetry directly if desired
```julia
  # viscous flux components in the x-direction
  f1 = zero(rho)
  f2 = tau_11 * nu
  f3 = tau_12 * nu
  f4 = ( v1 * tau_11 + v2 * tau_12 + q1 ) * nu

  # viscous flux components in y-direction
  g1 = zero(rho)
  g2 = f3 # tau_21 * nu
  g3 = tau_22 * nu
  # g4 = ( v1 * tau_21 + v2 * tau_22 + q2 ) * nu
  g4 = ( v1 * tau_12 + v2 * tau_22 + q2 ) * nu
```

Reply:
Resolving this conversation - we discussed using a `NGRADIENTS` field instead of `NVARIABLES` to distinguish parabolic gradient variables.